src/main/java/pl/gov/coi/pomocua/ads/configuration/ControllerAdviceConfiguration.java

```diff
@@ -1,29 +1,59 @@
 package pl.gov.coi.pomocua.ads.configuration;
 
+import lombok.RequiredArgsConstructor;
+import org.springframework.context.i18n.LocaleContextHolder;
+import org.springframework.core.annotation.AnnotatedElementUtils;
 import org.springframework.http.HttpStatus;
 import org.springframework.http.ResponseEntity;
 import org.springframework.validation.BindException;
 import org.springframework.validation.DataBinder;
 import org.springframework.web.bind.annotation.ControllerAdvice;
 import org.springframework.web.bind.annotation.ExceptionHandler;
 import org.springframework.web.bind.annotation.InitBinder;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import pl.gov.coi.pomocua.ads.error.CodeException;
+import pl.gov.coi.pomocua.ads.error.CodeValidationError;
 import pl.gov.coi.pomocua.ads.error.ValidationError;
-import pl.gov.coi.pomocua.ads.error.ValidationErrorResponse;
+import pl.gov.coi.pomocua.ads.error.ErrorResponse;
+import pl.gov.coi.pomocua.ads.mail.MessageProvider;
 
+import java.util.Optional;
+
+@RequiredArgsConstructor
 @ControllerAdvice
 public class ControllerAdviceConfiguration {
+
+    private final MessageProvider messageProvider;
+
     @InitBinder
     private void initDirectFieldAccess(DataBinder dataBinder) {
         dataBinder.initDirectFieldAccess();
     }
 
     @ExceptionHandler(value = {BindException.class})
-    public ResponseEntity<ValidationErrorResponse> handle(BindException e) {
+    public ResponseEntity<ErrorResponse> handle(BindException e) {
 
-        ValidationErrorResponse response = new ValidationErrorResponse(HttpStatus.BAD_REQUEST);
+        ErrorResponse response = new ErrorResponse(HttpStatus.BAD_REQUEST);
         e.getGlobalErrors().forEach(error -> response.addError(new ValidationError(error)));
         e.getFieldErrors().forEach(error -> response.addError(new ValidationError(error)));
 
         return ResponseEntity.status(response.getStatus()).body(response);
     }
+
+    @ExceptionHandler(value = {CodeException.class})
+    public ResponseEntity<ErrorResponse> handle(CodeException e) {
+
+        ErrorResponse response = new ErrorResponse(resolveHttpStatus(e));
+
+        CodeValidationError validationError = e.getValidationError();
+        String message = messageProvider.getMessageByCodeAndLocale(validationError.getMessageCode(), LocaleContextHolder.getLocale());
+        response.addError(new ValidationError(validationError.getField(), message, validationError.getType()));
+
+        return ResponseEntity.status(response.getStatus()).body(response);
+    }
+
+    private HttpStatus resolveHttpStatus(Exception e) {
```

Comment:
Wouldn't it be simpler if `CodeException`'s constructor just accepted a `HttpStatus` as an argument?

Reply:
```@ResponseStatus``` is also a proper way, this annotation was used from very beginning

Reply:
Yes I know `@ResponseStatus` is a proper way, but here it's not used as designed by Spring, but it's manually checked by our code (using reflection), so we're not using the ready Spring mechanism.
My idea was to pass HTTP status literally as a constructor argument, so that the reflection wouldn't be needed, as we're not using the Spring mechanism anyway :)

But nevermind, it doesn't affect the behavior, the code would be just slightly simpler.